src/game_events/manager.cpp

```diff
@@ -119,26 +121,65 @@ void manager::write_events(config& cfg) const
 
 void manager::execute_on_events(const std::string& event_id, manager::event_func_t func)
 {
-	// Start with a list of weak_ptrs to all matching event names (ie, "moveto").
-	handler_list matching_events = event_handlers_->get(event_id);
-
-	// Then check events with variables in their names for a match.
-	for(const auto& handler : event_handlers_->get_dynamic()) {
-		if(handler_ptr ptr = handler.lock()) {
-			if(ptr->matches_name(event_id, resources::gamedata)) {
-				matching_events.push_back(handler);
-			}
+	const std::string standardized_event_id = event_handlers::standardize_name(event_id);
+	const game_data* gd = resources::gamedata;
+	auto& active_handlers = event_handlers_->get_active();
+
+	// Save the end outside the loop so the end point remains constant,
+	// even if new events are added to the queue.
+	const unsigned saved_end = active_handlers.size();
+
+	// It's possible for this function to call itself again. If list cleanup occurred
+	// after each recursive call, the saved_end variable in the call above it would no
+	// longer be a valid end index. To work around that, we delay cleanup until the
+	// toplevel call is done running through the events (ie, when recursion_depth is 0).
+	static unsigned recursion_depth = 0;
+	++recursion_depth;
+
+	for(unsigned i = 0; i < saved_end; ++i) {
+		handler_ptr handler = nullptr;
+
+		try {
+			handler = active_handlers.at(i);
+		} catch(const std::out_of_range&) {
+			continue;
 		}
-	}
 
-	for(auto& handler : matching_events) {
-		if(handler_ptr ptr = handler.lock()) {
-			func(*this, ptr);
+		// Shouldn't happen, but we're just being safe.
+		if(!handler || handler->disabled()) {
+			continue;
+		}
+
+		// Could be more than one.
+		for(const std::string& name : handler->names()) {
+			bool matches = false;
+
+			if(utils::might_contain_variables(name)) {
+				// If we don't have gamedata, we can't interpolate variables, so there's
+				// no way the name will match. Move on to the next one in that case.
+				if(!gd) {
+					continue;
+				}
+
+				matches = standardized_event_id ==
+					event_handlers::standardize_name(utils::interpolate_variables_into_string(name, *gd));
```

Comment:
I think a `standardize_name` function is actually slightly _less_ efficient than a comparison that treats space and underscores as equal.

Reply:
thanks, I’m glad someone else noticed that... unfortunately it looks like the old code is being tossed out despite being more efficient in the general case. The new code could only be more efficient in some contrived example with hundreds of commas— in other words, never.

Reply:
I figured the difference in performance should be minimal enough as to be unnoticeable, but I haven't profiled to see.

Reply:
Well, you might be right, but the specialized comparison function has the additional benefit of not potentially breaking the event name when it contains a variable (if for whatever reason the variable hasn't yet been substituted). And in the worst case, the specialized comparison function runs through both strings once, while using the standardize_name runs through one of the strings twice. (An equality check runs through each string once; the standardize_name function runs through its argument once.) So, specialized comparison is 2n while standardize_name is 3n where n is the length of the string. Also, "worst-case" here is actually quite common, as it's the case where the names _are_ equal.

Reply:
I see. Perhaps I should restore the old comparison function.